docs/guide/widgets.md

```diff
@@ -388,3 +384,191 @@ The following builtin widgets use the Line API. If you are building advanced wid
+</div>
+
+In the following code we will implement the three widgets. There will be no functionality yet, but it should look like our design.
+
+=== "byte01.py"
+
+    ```python title="byte01.py" hl_lines="28-30 48-50 67-71"
+    --8<-- "docs/examples/guide/compound/byte01.py"
+    ```
+
+=== "Output"
+
+    ```{.textual path="docs/examples/guide/compound/byte01.py" columns="90" line="30"}
+    ```
+
+Note the `compose()` methods of each of the widgets.
+
+- The `BitSwitch` yields a [Label](../widgets/label.md) which displays the bit number, and a [Switch](../widgets/switch.md) control for that bit. The default CSS for `BitSwitch` aligns its children vertically, and sets the label's [text-align](../styles/text_align.md) to center.
+
+- The `ByteInput` yields 8 `BitSwitch` widgets and arranges them horizontally. It also adds a `focus-within` style in its CSS to draw an accent border when any of the switches are focused.
+
+- The `ByteEditor` yields a `ByteInput` and an Input control. The default CSS stacks the two controls on top of each other to divide the screen in to two parts.
+
+With these three widgets, the [DOM](CSS.md#the-dom) for our app will look like this:
+
+<div class="excalidraw">
+--8<-- "docs/images/byte_input_dom.excalidraw.svg"
+</div>
+
+Now that we have the design in place, we can implement the behavior.
+
+
+### Data flow
+
+We want to ensure that our widgets are re-usable, which we can do by following the guideline of "attributes down, messages up". This means that a widget can update a child by setting its attributes or calling its methods, but widgets should only ever send [messages](./events.md) to their *parent*.
+
+!!! info
+
+    This pattern of only setting attributes in one direction and using messages for the opposite direction is known as *uni-directional data flow*.
+
+In practice, this means that to update a child widget you get a reference to it and use it like any other Python object. Here's an example of an [action](actions.md) that updates a child widget:
+
+```python
+def action_set_true(self):
+    self.query_one(Switch).value = 1
+```
+
+If a child needs to update a parent, it should send a message with [post_message][textual.message_pump.MessagePump.post_message].
+
+Here's an example of posting message:
+
+```python
+def on_click(self):
+    self.post_message(MyWidget.Change(active=True))
+```
+
+Note that *attributes down and messages up* means that you can't modify widgets on the same level directly. If you want to modify a *sibling*, you will need to send a message to the parent, and the parent would make the changes.
+
+The following diagram illustrates this concept:
+
+
+<div class="excalidraw">
+--8<-- "docs/images/attributes_messages.excalidraw.svg"
+</div>
+
+### Messages up
+
+Let's extend the ByteEditor so that clicking any of the 8 `BitSwitch` widgets updates the decimal value. To do this we will add a custom message to `BitSwitch` that we catch in the `ByteEditor`.
```

Comment:
```suggestion
Let's extend the `ByteEditor` so that clicking any of the 8 `BitSwitch` widgets updates the decimal value. To do this we will add a custom message to `BitSwitch` that we catch in the `ByteEditor`.
```